File: config/core/controllers/my_farm_controller.py
"""
KisanAI OS
My Farm Controller
Version: 1.0.0
"""

from config.core.models.user import User
from config.core.services.my_farm_service import MyFarmService


class MyFarmController:
    """My Farm Controller - self-service farm management."""

    def __init__(self, session=None):
        self.service = MyFarmService(session)

    def get_farm(self, user: User):
        return self.service.get_farm(user.id)

    def create_farm(self, user: User, farm_data):
        return self.service.create_farm(user, farm_data)

    def update_farm(self, user: User, farm_data):
        return self.service.update_farm(user.id, farm_data)

    def delete_farm(self, user: User):
        return self.service.delete_farm(user.id)

    def get_crops(self, user: User):
        return self.service.get_crops(user.id)

    def add_crop(self, user: User, crop_data):
        return self.service.add_crop(user.id, crop_data)

    def update_crop(self, user: User, crop_id, crop_data):
        return self.service.update_crop(user.id, crop_id, crop_data)

    def delete_crop(self, user: User, crop_id):
        return self.service.delete_crop(user.id, crop_id)

    def close(self):
        self.service.close()

File: mobile/lib/features/my_farm/my_farm_controller.dart
import 'package:flutter/foundation.dart';

import '../../core/constants/app_strings.dart';
import '../../core/errors/api_exception.dart';
import '../../core/errors/error_messages.dart';
import '../../models/crop.dart';
import '../../models/farmer.dart';
import '../../services/my_farm_api.dart';

/// Load state of a [MyFarmController].
enum MyFarmState { initial, loading, ready, error }

/// Owns the current user's farm: profile fields plus planted crops.
///
/// `farm` is null until the user creates one; the UI then shows the create
/// form. Write operations return `true` on success and refresh the loaded
/// data; failures are surfaced through [errorMessage].
class MyFarmController extends ChangeNotifier {
  MyFarmController(this._api);

  final MyFarmApi _api;

  MyFarmState state = MyFarmState.initial;
  Farmer? farm;
  List<Crop> crops = const [];
  String? errorMessage;
  bool busy = false;

  bool get hasFarm => farm != null;

  Future<void> load() async {
    state = MyFarmState.loading;
    errorMessage = null;
    notifyListeners();
    try {
      farm = await _api.fetchFarm();
      crops = farm == null ? const [] : await _api.fetchCrops();
      state = MyFarmState.ready;
    } on ApiException catch (e) {
      errorMessage = errorMessageFor(e);
      state = MyFarmState.error;
    } catch (_) {
      errorMessage = AppStrings.genericError;
      state = MyFarmState.error;
    }
    notifyListeners();
  }

  Future<bool> createFarm({
    required String village,
    required String district,
    required String state,
    double? farmSize,
  }) {
    return _save(
      () => _api.createFarm(
        village: village,
        district: district,
        state: state,
        farmSize: farmSize,
      ),
      reload: true,
    );
  }

  Future<bool> updateFarm({
    required String village,
    required String district,
    required String state,
    double? farmSize,
  }) {
    return _save(
      () => _api.updateFarm(
        village: village,
        district: district,
        state: state,
        farmSize: farmSize,
      ),
      reload: true,
    );
  }

  Future<bool> deleteFarm() {
    return _save(_api.deleteFarm, reload: true);
  }

  Future<bool> addCrop({
    required String cropName,
    required String season,
    required int durationDays,
    required String waterRequirement,
  }) {
    return _save(
      () => _api.addCrop(
        cropName: cropName,
        season: season,
        durationDays: durationDays,
        waterRequirement: waterRequirement,
      ),
      reload: true,
    );
  }

  Future<bool> updateCrop(
    Crop crop, {
    required String cropName,
    required String season,
    required int durationDays,
    required String waterRequirement,
  }) {
    return _save(
      () => _api.updateCrop(
        crop.cropId,
        cropName: cropName,
        season: season,
        durationDays: durationDays,
        waterRequirement: waterRequirement,
      ),
      reload: true,
    );
  }

  Future<bool> deleteCrop(Crop crop) {
    return _save(() => _api.deleteCrop(crop.cropId), reload: true);
  }

  Future<bool> _save(
    Future<void> Function() action, {
    bool reload = false,
  }) async {
    busy = true;
    errorMessage = null;
    notifyListeners();
    try {
      await action();
      if (reload) {
        await load();
      }
      return true;
    } on ApiException catch (e) {
      errorMessage = errorMessageFor(e);
      return false;
    } catch (_) {
      errorMessage = AppStrings.genericError;
      return false;
    } finally {
      busy = false;
      notifyListeners();
    }
  }
}
